db-queue-spring/src/main/java/ru/yoomoney/tech/dbqueue/spring/dao/H2QueuePickTaskDao.java

```diff
@@ -0,0 +1,219 @@
+                            .withNextProcessAt(getZonedDateTime(rs, queueTableSchema.getNextProcessAtField()))
+                            .withPayload(rs.getString(queueTableSchema.getPayloadField()))
+                            .withAttemptsCount(rs.getLong(queueTableSchema.getAttemptField()))
+                            .withReenqueueAttemptsCount(rs.getLong(queueTableSchema.getReenqueueAttemptField()))
+                            .withTotalAttemptsCount(rs.getLong(queueTableSchema.getTotalAttemptField()))
+                            .withExtData(additionalData).build();
+
+                }, location.getQueueId().asString(), pickTaskSettings.getRetryInterval().getSeconds());
+    }
+
+    public static class PickerProcedure {
+        private static final Map<String, PickQuery> pickTaskSqlCache = new ConcurrentHashMap<>();
+
+        public static ResultSet pickTask(final Connection connection,
+                                         final String queueName,
+                                         final long retryInterval) throws SQLException {
+
+            PickQuery pickQuery = pickTaskSqlCache.get(queueName);
+            Objects.requireNonNull(pickQuery, "A pick query can't be null");
+
+            PreparedStatement chooseStatement = connection.prepareStatement(pickQuery.selectSql);
+            chooseStatement.setString(1, queueName);
+            ResultSet chooseResultSet = chooseStatement.executeQuery();
+            if (!chooseResultSet.next())
+                return chooseResultSet;
+
+            final long taskId = chooseResultSet.getLong(1);
+            PreparedStatement updateStatement = connection.prepareStatement(pickQuery.updateSql);
+            updateStatement.setLong(1, retryInterval);
+            updateStatement.setLong(2, taskId);
+            int updatedRowCount = updateStatement.executeUpdate();
+            if (updatedRowCount != 1)
+                throw new IllegalStateException("Something wrong went here. Only row must be updated, not more!");
+
+            final PreparedStatement selectStatement = connection.prepareStatement(pickQuery.returnSql);
+            selectStatement.setLong(1, taskId);
+            return selectStatement.executeQuery();
+        }
+
+        private static String createPickQuery(QueueLocation location,
+                                              PickTaskSettings pickTaskSettings,
+                                              QueueTableSchema queueTableSchema) {
+
+            pickTaskSqlCache.computeIfAbsent(location.getQueueId().asString(), ignoredKey -> {
+
+                final String selectSql = String.format("" +
+                                "SELECT %s " +
+                                "FROM %s " +
+                                "WHERE %s = ? " +
+                                "  AND %s <= now() " +
+                                "ORDER BY %s ASC " +
+                                "LIMIT 1 " +
+                                "FOR UPDATE",
```

Comment:
Does h2 have a skip-locked feature? What is the consequence in case we don't skip locked rows? I suppose db-queue performance will degrade to single-thread processing. Is that right?

Reply:
By now, the H2 database doesn't support this feature, moreover, it is not going to be implemented in the nearest future.             
I have made up a solution based on ConcurrentHashMap. Yet, I am not sure about performance, I suppose, it can be much better.